deeppavlov/models/ranking/siamese_predictor.py

```diff
@@ -0,0 +1,137 @@
+
+from deeppavlov.core.common.log import get_logger
+from deeppavlov.core.models.component import Component
+from deeppavlov.models.ranking.keras_siamese_model import SiameseModel
+from deeppavlov.core.data.simple_vocab import SimpleVocabulary
+from deeppavlov.core.common.registry import register
+
+log = get_logger(__name__)
+
+@register('siamese_predictor')
+class SiamesePredictor(Component):
+    """The class for ranking or paraphrase identification using the trained siamese network  in the ``interact`` mode.
+
+    Args:
+        batch_size: A size of a batch.
+        num_context_turns: A number of ``context`` turns in data samples.
+        ranking: Whether to perform ranking.
+            If it is set to ``False`` paraphrase identification will be performed.
+        attention: Whether any attention mechanism is used in the siamese network.
+            If ``False`` then calculated in advance vectors of ``responses``
+            will be used to obtain similarity score for the input ``context``;
+            Otherwise the whole siamese architecture will be used
+            to obtain similarity score for the input ``context`` and each particular ``response``.
+            The parameter will be used if the ``ranking`` is set to ``True``.
+        responses: A instance of :class:`~deeppavlov.core.data.simple_vocab.SimpleVocabulary`
+            with all possible ``responses`` to perform ranking.
+            Will be used if the ``ranking`` is set to ``True``.
+        preproc_func: A ``__call__`` function of the
+            :class:`~deeppavlov.models.preprocessors.siamese_preprocessor.SiamesePreprocessor`.
+        interact_pred_num: The number of the most relevant ``responses`` which will be returned.
+            Will be used if the ``ranking`` is set to ``True``.
+        **kwargs: Other parameters.
+    """
+
+    def __init__(self,
+                 model: SiameseModel,
+                 batch_size: int,
+                 num_context_turns: int = 1,
+                 ranking: bool = True,
+                 attention: bool = False,
+                 responses: SimpleVocabulary = None,
+                 preproc_func: Callable = None,
+                 interact_pred_num: int = 3,
+                 **kwargs):
+
+        super().__init__()
+
+        self.batch_size = batch_size
+        self.num_context_turns = num_context_turns
+        self.ranking = ranking
+        self.attention = attention
+        self.responses = {el[1]: el[0] for el in responses.items()}
+        self.preproc_responses = []
+        self.response_embeddings = None
+        self.preproc_func = preproc_func
+        self.interact_pred_num = interact_pred_num
+        self.model = model
+        self._build_preproc_responses()
+        if not self.attention:
+            self._build_response_embeddings()
+
+    def __call__(self, batch: Iterable[List[np.ndarray]]) -> List[Union[List[str],str]]:
+        context = next(batch)
```

Comment:
I think if this component can only work with batches of size 1, it should rise an error if the batch is bigger

Reply:
corrected